### PR TITLE
Add pretraining for CNN models

### PR DESCRIPTION
**Pull request by daemon**
- Pretrain on CIFAR-10, CIFAR-100, and FashionMNIST

### CODE REVIEW

File: ww4ff/run/pretrain.py [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+                                    transforms.ToTensor(),
+                                    transforms.Normalize((0.5,), (0.5,)),
+                                    expand])
+    test_transform = transforms.Compose([transforms.Pad((2, 2)),
+                                         transforms.ToTensor(),
+                                         transforms.Normalize((0.5,), (0.5,)),
+                                         expand])
+    trainset3 = torchvision.datasets.FashionMNIST(root='./data', train=True, download=True, transform=transform)
+    testset3 = torchvision.datasets.FashionMNIST(root='./data', train=False, download=True, transform=test_transform)
+    train_dl = tud.DataLoader(tud.ConcatDataset([trainset1, trainset2, trainset3]),
```
Comment (ljj7975) [inline ww4ff/run/pretrain.py:72]: isn't CIFAR100 has 100 classes? How can you concatenate a dataset of 10 classes with a dataset of 100 classes?

File: ww4ff/run/train.py [hunk 2]
```diff
@@ -10,93 +9,105 @@
...
-        batch.to(device)
-        zmuv_transform.update(std_transform(batch.audio_data))
-
+    if (ws.path / 'zmuv.pt.bin').exists():
+        zmuv_transform.load_state_dict(torch.load(str(ws.path / 'zmuv.pt.bin')))
+    else:
+        for idx, batch in enumerate(tqdm(prep_dl, desc='Constructing ZMUV')):
+            batch.to(device)
+            zmuv_transform.update(std_transform(batch.audio_data))
+            if idx == 2000:  # TODO: quick debugging, remove later
```
Comment (ljj7975) [inline ww4ff/run/train.py:95]: should we keep it still?

File: ww4ff/run/train.py [hunk 3]
```diff
@@ -10,93 +9,105 @@
...
     print_stats('Wake word dataset', ww_train_ds, ww_dev_ds, ww_test_ds)
 
     sr = SETTINGS.audio.sample_rate
     wind_sz = int(SETTINGS.training.eval_window_size_seconds * sr)
     stri_sz = int(SETTINGS.training.eval_stride_size_seconds * sr)
-    ww_train_ds = WakeWordTrainingDataset(ww_train_ds, ww)
-    ww_dev_ds = WakeWordEvaluationDataset(WakeWordTrainingDataset(ww_dev_ds, ww), wind_sz, stri_sz)
-    ww_test_ds = WakeWordEvaluationDataset(WakeWordTrainingDataset(ww_test_ds, ww), wind_sz, stri_sz)
+    ww_dev_pos_ds = ww_dev_ds.filter(lambda x: x.compute_frame_labels(args.vocab), clone=True)
+    ww_dev_neg_ds = ww_dev_ds.filter(lambda x: not x.compute_frame_labels(args.vocab), clone=True)
```
Comment (ljj7975) [inline ww4ff/run/train.py:64]: if I understood correctly, "hey xxxx" will still fall into pos case as compute_frame_labels non-empty dict.
Can you double check?

File: ww4ff/data/dataset/dataset.py [hunk 4]
```diff
@@ -138,9 +138,24 @@ def _inc(self):
...
+        def emit_next_positive_example(self):
+            example = self.dataset.dataset[self.curr_file_idx]
+            pos_ts, pos_lbl = self.frame_labels[self.label_pointer]
+            self.output_positives.append(pos_ts)
+            b = int(pos_ts / 1000 * self.dataset.dataset.sr)
+            a = max(b - self.dataset.window_size, 0)
+            self.label_pointer += 1
+            return ClassificationBatch.from_single(example.audio_data[..., a:b], pos_lbl)
+
         def __next__(self) -> ClassificationBatch:
```
Comment (ljj7975) [inline ww4ff/data/dataset/dataset.py:150]: Can you explain this next code little?
it's hard to follow and understand

File: ww4ff/run/pretrain.py (hunk 1 above)
Comment (daemon) [reply ww4ff/run/pretrain.py:72]: I just did mod 10.

File: ww4ff/run/train.py (hunk 2 above)
Comment (daemon) [reply ww4ff/run/train.py:110]: Yeah, just for speed. Also, CLT says 2000 should be fine.

File: ww4ff/run/train.py (hunk 3 above)
Comment (daemon) [reply ww4ff/run/train.py:79]: Yup, that's correct. For the negative set, we want speech where there are no positive labels at all.

File: ww4ff/data/dataset/dataset.py (hunk 4 above)
Comment (daemon) [reply ww4ff/data/dataset/dataset.py:150]: Yeah, I should document everything at some point. For evaluation, we stride a window (from left to right) across the entire dataset, concatenating files as we go and outputting an example with the appropriate label for each window.
